pkg/controller/keyvault/gateway_secret_provider_class.go

```diff
@@ -0,0 +1,232 @@
+)
+
+// GatewaySecretProviderClassReconciler manages a SecretProviderClass for Gateway resource that specifies a ServiceAccount
+// and Keyvault URI in its TLS options field. The SPC is used to mirror the Keyvault values into
+// a k8s secret so that it can be used by the CRD controller.
+type GatewaySecretProviderClassReconciler struct {
+	client client.Client
+	events record.EventRecorder
+	config *config.Config
+}
+
+func NewGatewaySecretClassProviderReconciler(manager ctrl.Manager, conf *config.Config) error {
+	metrics.InitControllerMetrics(gatewaySecretProviderControllerName)
+
+	return gatewaySecretProviderControllerName.AddToController(
+		ctrl.
+			NewControllerManagedBy(manager).
+			For(&gatewayv1.Gateway{}).
+			Owns(&secv1.SecretProviderClass{}), manager.GetLogger(),
+	).Complete(&GatewaySecretProviderClassReconciler{
+		client: manager.GetClient(),
+		events: manager.GetEventRecorderFor("aks-app-routing-operator"),
+		config: conf,
+	})
+}
+
+func (g *GatewaySecretProviderClassReconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, retErr error) {
+	// set up metrics given result/error
+	defer func() {
+		metrics.HandleControllerReconcileMetrics(gatewaySecretProviderControllerName, res, retErr)
+	}()
+
+	// set up logger
+	logger, err := logr.FromContext(ctx)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("creating logger: %w", err)
+	}
+	logger = gatewaySecretProviderControllerName.AddToLogger(logger).WithValues("name", req.Name, "namespace", req.Namespace)
+
+	// retrieve gateway resource from request + log the get attempt, but ignore not found
+	gwObj := &gatewayv1.Gateway{}
+	err = g.client.Get(ctx, req.NamespacedName, gwObj)
+	if err != nil {
+		if client.IgnoreNotFound(err) != nil {
+			logger.Error(err, "failed to fetch Gateway")
+			return ctrl.Result{}, fmt.Errorf("fetching gateway: %w", err)
+		}
+		return ctrl.Result{}, nil
+	}
+
+	if !shouldReconcileGateway(gwObj) {
+		return ctrl.Result{}, nil
+	}
+
+	// check its TLS options - needs to have both cert uri and either serviceaccount name or clientid
+	for index, listener := range gwObj.Spec.Listeners {
+		spc := &secv1.SecretProviderClass{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "secrets-store.csi.x-k8s.io/v1",
+				Kind:       "SecretProviderClass",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      generateGwListenerCertName(gwObj.Name, listener.Name),
+				Namespace: req.Namespace,
+				Labels:    manifests.GetTopLevelLabels(),
+				OwnerReferences: []metav1.OwnerReference{{
+					APIVersion: gwObj.APIVersion,
+					Controller: util.ToPtr(true),
+					Kind:       gwObj.Kind,
+					Name:       gwObj.Name,
+					UID:        gwObj.UID,
+				}},
+			},
+		}
+		logger = logger.WithValues("spc", spc.Name)
+
+		if listenerIsKvEnabled(listener) {
+			var clientId string
+			clientId, err = retrieveClientIdFromListener(ctx, g.client, req.Namespace, listener.TLS.Options)
```

Comment:
the client id should come from the SA.

We need some way of triggering a reconcile if the SA changes their client id too

Reply:
done via sa indexer